### PR TITLE
CryptoMB - TLS handshake acceleration for Istio

### PR DESCRIPTION
**Pull request by rveerama1**
Blogpost for TLS handshake acceleration for Istio using CryptoMB Private Key Provider.

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

- [ ] Configuration Infrastructure
- [X] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by craigbox** [comment]
Preview is [here](https://deploy-preview-11302--preliminary-istio.netlify.app/latest/blog/2022/cryptomb-privatekeyprovider/).  Review might be slower due to KubeCon next week; did you have a date you wanted to publish?

Initial thought: very cool!

**Comment by rveerama1** [comment]
> Preview is [here](https://deploy-preview-11302--preliminary-istio.netlify.app/latest/blog/2022/cryptomb-privatekeyprovider/). Review might be slower due to KubeCon next week; did you have a date you wanted to publish?
> 
> Initial thought: very cool!

Thank you. We are going to talk about this topic in Kubecon this week. It would be nice if it is available for users soon.

**Comment by irisdingbj** [comment]
LGTM

File: content/en/blog/2022/cryptomb-privatekeyprovider/index.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+attribution: "Ravi kumar Veeramally (Intel), Ismo Puustinen (Intel), Sakari Poussa (Intel)"
+keywords: [Istio, CryptoMB, gateways, sidecar]
+---
+
+Cryptographic operations are among the most compute-intensive and critical operations when it comes to secured connections. Istio uses Envoy as the "gateways/sidecar" to handle secure connections and intercept the traffic.
+
+Depending upon use cases, when an ingress gateway must handle a lot of incoming TLS connections and secured connections between service-to-service through sidecar proxies, the load on Envoy increases. The potential performance depends on many factors, such as size of the cpuset on which Envoy is running, incoming traffic patterns, and key size. These factors can impact Envoy serving many new incoming TLS requests. To achieve performance improvements and accelerated handshakes, a new feature was introduced in Envoy 1.20 and Istio 1.14. It can be achieved with 3rd Gen Intel® Xeon® Scalable processors, the Intel® Integrated Performance Primitives (Intel® IPP) [crypto library], CryptoMB Private Key Provider Method support in Envoy, and Private Key Provider configuration in Istio using ProxyConfig.
```
Comment (ericvn) [inline content/en/blog/2022/cryptomb-privatekeyprovider/index.md:11]: - TLS and secured service-to-service connections thorough
- Not sure I like `a lot`. Maybe `a large number` is better. 
- [crypto library] - remove the []'s?

File: content/en/blog/2022/cryptomb-privatekeyprovider/index.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+{{< image link="./envoy-boringssl-pkp-flow.png" caption="Envoy <-> BoringSSL <-> PrivateKeyProvider" >}}
+
+The Envoy worker thread has a buffer size for eight RSA requests. When the first RSA request is stored in the buffer, a timer will be initiated (timer duration is set by the poll_delay field in the CryptoMB configuration).
+
+{{< image link="./timer-started.png" caption="Buffer timer started" >}}
+
+When the buffer is full or when the timer expires, perform the crypto operations for all RSA requests simultaneously. The SIMD (single instruction, multiple data) processing gives the potential performance benefit compared to the non-accelerated case.
```
Comment (ericvn) [inline content/en/blog/2022/cryptomb-privatekeyprovider/index.md:31]: c/perform the crypto operations/the crypto operations are performed

File: content/en/blog/2022/cryptomb-privatekeyprovider/index.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+{{< image link="./timer-expired.png" caption="Buffer timer expired" >}}
+
+## Envoy CryptoMB Private Key Provider configuration
+
+A regular TLS configuration uses just a private key. When a private key provider is used, the private key field is replaced with a private key provider field. It again contains two fields, provider name and typed config. Typed config is CryptoMbPrivateKeyMethodConfig, and it specifies the private key and the poll delay.
```
Comment (ericvn) [inline content/en/blog/2022/cryptomb-privatekeyprovider/index.md:37]: - only uses
- It ~~again~~ contains

File: content/en/blog/2022/cryptomb-privatekeyprovider/index.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    provider_name: cryptomb
+    typed_config:
+      "@type": type.googleapis.com/envoy.extensions.private_key_providers.cryptomb.v3alpha.CryptoMbPrivateKeyMethodConfig
+      private_key: { "filename": "/path/key.pem" }
+      poll_delay: 10ms
+{{< /text >}}
+
+## Istio CryptoMB Private Key Provider configuration
+
+In Istio, CryptoMB private key provider configuration can be applied to mesh wide, gateways specific or pod specific configurations using pod annotations. The User will provide the PrivateKeyProvider in the ProxyConfig with the pollDelay value. This configuration will be applied to mesh wide (gateways and all sidecars).
```
Comment (ericvn) [inline content/en/blog/2022/cryptomb-privatekeyprovider/index.md:62]: applied ~~to~~ mesh wide

File: content/en/blog/2022/cryptomb-privatekeyprovider/index.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        name: httpbin
+        ports:
+        - containerPort: 80
+{{< /text >}}
+
+### Performance
+
+The potential performance benefit depends on many factors. For example, the size of the cpuset Envoy is running on, incoming traffic pattern, encryption type (RSA or ECDSA), and key size. There is a white paper with [performance](https://www.intel.com/content/www/us/en/architecture-and-technology/crypto-acceleration-in-xeon-scalable-processors-wp.html) numbers.
+
+Istio ingress gateway TLS handshake performance comparison on 1.14-dev on May10th 2022.
```
Comment (ericvn) [inline content/en/blog/2022/cryptomb-privatekeyprovider/index.md:176]: May 10th,

**Comment by rveerama1** [comment]
@ericvn Thanks for the review, fixed as per your suggestions. PTAL.

**Comment by rveerama1** [comment]
@ericvn @craigbox PTAL.

**Comment by ericvn** [comment]
LGTM, but will let @craigbox have the final look.

**Comment by rveerama1** [comment]
ping @craigbox

**Review by rvennam — APPROVED** [review]

**Comment by rveerama1** [comment]
@craigbox when you added cherrypick/release-1.14 label, does it mean it should be here https://github.com/istio/istio.io/commits/release-1.14 right? May I know when will it publish here https://istio.io/latest/blog/ .

File: content/en/blog/2022/cryptomb-privatekeyprovider/index.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        - containerPort: 80
+{{< /text >}}
+
+### Performance
+
+The potential performance benefit depends on many factors. For example, the size of the cpuset Envoy is running on, incoming traffic pattern, encryption type (RSA or ECDSA), and key size. Here is a white paper with [performance](https://www.intel.com/content/www/us/en/architecture-and-technology/crypto-acceleration-in-xeon-scalable-processors-wp.html) numbers.
+
+Istio ingress gateway TLS handshake performance comparison on 1.14-dev on May 10th 2022.
+
+{{< image link="./istio-ingress-gateway-tls-handshake-perf-num.png" caption="Istio ingress gateway TLS handshake performance comparison" >}}
```
Comment (howardjohn) [inline content/en/blog/2022/cryptomb-privatekeyprovider/index.md:178]: 40ms is extremely higher than expected (order of magnitude). Are you sure the test was configured correctly? Do you have more details on the setup?

Is most of the time spent on network latency from k6 -> cluster?

Comment (irisdingbj) [reply content/en/blog/2022/cryptomb-privatekeyprovider/index.md:178]: yes, this is a total latency including k6- > gateway- >fortio server, so it is higher, We will submit a PR to clearly state this out and our detailed settings.

**Pull request merged**